hathor/indexes/rocksdb_mempool_tips_index.py

```diff
@@ -0,0 +1,134 @@
+        assert tx.hash is not None
+        assert tx.storage is not None
+        to_remove: Set[bytes] = set()
+        to_remove_parents: Set[bytes] = set()
+        tx_storage = tx.storage
+        for tip_tx in self.iter(tx_storage):
+            assert tip_tx.hash is not None
+            # A new tx/block added might cause a tx in the tips to become voided. For instance,
+            # there might be twin txs, tx1 and tx2, where tx1 is valid and tx2 voided. A new block
+            # confirming tx2 will make it valid while tx1 becomes voided, so it has to be removed
+            # from the tips. The txs confirmed by tx1 need to be double checked, as they might
+            # themselves become tips (hence we use to_remove_parents)
+            meta = tip_tx.get_metadata()
+            if meta.voided_by:
+                to_remove.add(tip_tx.hash)
+                to_remove_parents.update(tip_tx.parents)
+                continue
+
+            # might also happen that a tip has a child that became valid, so it's not a tip anymore
+            confirmed = False
+            for child_meta in map(tx_storage.get_metadata, meta.children):
+                assert child_meta is not None
+                if not child_meta.voided_by:
+                    confirmed = True
+                    break
+            if confirmed:
+                to_remove.add(tip_tx.hash)
+
+        if to_remove:
+            self._remove_many(to_remove)
+            self.log.debug('removed voided txs from tips', txs=[tx.hex() for tx in to_remove])
+
+        # Check if any of the txs being confirmed by the voided txs is a tip again. This happens
+        # if it doesn't have any other valid child.
+        to_add = set()
+        for tx_hash in to_remove_parents:
+            confirmed = False
+            # check if it has any valid children
+            meta = not_none(tx_storage.get_metadata(tx_hash))
+            if meta.voided_by:
+                continue
+            children = meta.children
+            for child_meta in map(tx_storage.get_metadata, children):
+                assert child_meta is not None
+                if not child_meta.voided_by:
+                    confirmed = True
+                    break
+            if not confirmed:
+                to_add.add(tx_hash)
+
+        if to_add:
+            self._index.update(to_add)
+            self.log.debug('added txs to tips', txs=[tx.hex() for tx in to_add])
+
+        if tx.get_metadata().voided_by:
+            # this tx is voided, don't need to update the tips
+            self.log.debug('voided tx, won\'t add it as a tip', tx=tx.hash_hex)
+            return
+
+        self._remove_many(set(tx.parents))
+
+        if tx.is_transaction and tx.get_metadata().first_block is None:
+            assert tx.hash is not None
+            self._index.add(tx.hash)
+
+    def iter(self, tx_storage: 'TransactionStorage', max_timestamp: Optional[float] = None) -> Iterator[Transaction]:
```

Comment:
What do you mean? Is it `iter` and `iter_all` being almost the same? They seem pretty different to me. Do you mean to refactor them into `RocksDBSimpleSet`? I'd also like to avoid giving meaning to the key/value of `RocksDBSimpleSet`, the way it is right now is very generic and can be used by other implementations without them worrying what to store.

Reply:
Actually I might have written not very well the comment. I was talking about the methods `iter` and `iter_all` from the two classes, rocksdb and memory. They are pretty much the same, just like the `update` method.

Reply:
Oh, I got it now. I think it's important to note that these methods are only identical because `RocksDBSimpleSet` behaves a lot like a `Set`. I have a good idea on how to make it more reusable.

Reply:
Done.

Reply:
I really like the way you reused the methods, I think it's much more clean now